### PR TITLE
feat: Add OpenTelemetry integration

### LINKED ISSUE #124: PubSub add OpenTelemetry support
OpenTelemetry should be added to PubSub to provide OpenTracing and OpenCensus functionality as one API and SDK. It would be helpful to use it to trace a message throughout its life cycle.

### PR DESCRIPTION
**Pull request by sethmaxwl**
Closes #124 
Requires. #158.

This PR adds OpenTelemetry tracing to publisher and subscriber clients.
This PR does not interfere with any underlying client functions. OpenTelemetry is an optional dependency that creates a trace that tracks a message from when it is published to when it is received by a subscriber.

### Ongoing concerns
- [x] Adding OpenTelemetry to unit tests as external dependencies.

### PR checklist
- [x]  Make sure to open an issue as a bug/issue before writing your code! That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by plamut** [comment]
This requires dropping Python 2 support, right? (#131)

**Comment by sethmaxwl** [comment]
> This requires dropping Python 2 support, right? (#131)

Yes, OpenTelemetry does not support any Python version before 3.4.

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:41]: (nit) `pass` is redundant here

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:50]: (nit)
Following the `mypy` notation, optional types are specified as `Optional[typename]`

Applies to multiple places.

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:46]: (nit)
Docstring summary lines should use imperative mood.

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:53]: A word seems to be missing in this sentence?

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:76]: Since `parent` is a mapping, it can be unpacked directly:
```py
parent_span_context = SpanContext(**parent)
```

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (plamut) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:78]: What exceptions are we catching here? It's advisable to be as specific as possible, since as written, the code will silently swallow even typos in variable names and similar unrelated errors.

File: google/cloud/pubsub_v1/publisher/client.py
Comment (plamut) [inline google/cloud/pubsub_v1/publisher/client.py:380]: Sanity check - what if somebody explicitly sets a message attribute named `googclient_OpenTelemetrySpanContext`?

It's probably quite unlikely, but it probably makes sense to at least issue a warning that this message attribute will be overridden.

File: google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py
Comment (plamut) [inline google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:626]: Dicts support direct containment check, i.e. `"foo" in my_dict`, we do not need to obtain the keys first with the `keys()` method.

File: noxfile.py
Comment (plamut) [inline noxfile.py:79]: This change will get wiped out on the next code regen, let's add dependencies through the `synth.py` file instead to make them persistent.

File: tests/unit/pubsub_v1/opentelemetry/test_opentelemetry.py
Comment (plamut) [inline tests/unit/pubsub_v1/opentelemetry/test_opentelemetry.py:60]: Recommendation - most (all?) of the existing unit tests use the `pytest` framework, thus let's be consistent and use it in these tests, too. Should make them more concise with less boilerplate. :wink:

File: tests/unit/pubsub_v1/publisher/test_publisher_client.py
Comment (plamut) [inline tests/unit/pubsub_v1/publisher/test_publisher_client.py:35]: Caution - this modifies the global state, i.e. the `sys.modules` dictionary, and the change is not restored after the publisher tests are run. Let's make sure that the fixture is undone when not needed anymore.

**Review by plamut — COMMENTED** [review]
Disclaimer - I would have to familiarize myself with the `opentelemetry` library more, thus I primarily focused on the code aspects, and less on the actual semantics. Added suggestions at places where the code could benefit from them.

File: noxfile.py
Comment (sethmaxwl) [reply noxfile.py:79]: Adding these as external dependencies may not work since there is not a synthtool option to add external unit test dependencies, although there is an option for external system test dependencies. Should these tests be added as system tests?

File: tests/unit/pubsub_v1/opentelemetry/test_opentelemetry.py
Comment (sethmaxwl) [reply tests/unit/pubsub_v1/opentelemetry/test_opentelemetry.py:60]: I used the unittest framework here because of the setup and teardown methods that make it easier to create the OpenTelemetry environment for the test cases. This can be done for each test in the `pytest` framework, but there would be some repeated code.

Comment (plamut) [reply tests/unit/pubsub_v1/opentelemetry/test_opentelemetry.py:60]: I'm curious where `@pytest.fixture` falls short here? In should be possible to achieve the same setup / teardown behavior without code repetition and just pass the fixture as an argument to the tests that need it.

I'm not super strongly opinionated here, though, the unittest framework does the job fine, it's just that using a single test framework means somewhat lower maintenance cost in the long run.

File: noxfile.py
Comment (plamut) [reply noxfile.py:79]: It's not ideal, but we can circumvent this limitation and add additional dependencies through the `synth.py` file. That way the changes to `noxfile.py` will not get lost on the next code regen.

Applying the following patch should do the trick:

```patch
diff --git noxfile.py noxfile.py
index 3af3d67..20b5ad2 100644
--- noxfile.py
+++ noxfile.py
@@ -70,14 +70,12 @@ def lint_setup_py(session):
 
 def default(session):
     # Install all test dependencies, then install this package in-place.
+
+    session.install("mock", "pytest", "pytest-cov")
     session.install(
-        "mock",
-        "pytest",
-        "pytest-cov",
-        "opentelemetry-api",
-        "opentelemetry-sdk",
-        "opentelemetry-instrumentation",
+        "opentelemetry-api", "opentelemetry-sdk", "opentelemetry-instrumentation",
     )
+
     session.install("-e", ".")
 
     # Run py.test against the unit tests.
diff --git synth.py synth.py
index 0e2c96e..10edae8 100644
--- synth.py
+++ synth.py
@@ -279,4 +279,18 @@ s.move(templated_files)
 # ----------------------------------------------------------------------------
 python.py_samples()
 
+# ----------------------------------------------------------------------------
+# Additional unit test dependencies
+# ----------------------------------------------------------------------------
+s.replace(
+    "noxfile.py",
+    r'session\.install\("mock", "pytest", "pytest-cov"\)',
+    """
+    \g<0>
+    session.install(
+        "opentelemetry-api", "opentelemetry-sdk", "opentelemetry-instrumentation",
+    )
+    """,
+)
+
 s.shell.run(["nox", "-s", "blacken"], hide_output=False)
```

**Comment by sethmaxwl** [comment]
> Disclaimer - I would have to familiarize myself with the `opentelemetry` library more, thus I primarily focused on the code aspects, and less on the actual semantics. Added suggestions at places where the code could benefit from them.

All suggestions have been addressed.

**Review by plamut — COMMENTED** [review]
LGTM now, although we might have to do some additional changes once [#158](https://github.com/googleapis/python-pubsub/pull/158) is merged.

If you have time, you might want to rebase your branch onto that PR branch and see if things work there, too.

File: google/cloud/pubsub_v1/publisher/client.py
Comment (plamut) [inline google/cloud/pubsub_v1/publisher/client.py:384]: Oh, forgot about this - I don't think we test this code path yet, i.e. whether a warning is indeed emitted.

File: README.rst
Comment (AndrewAXue) [inline README.rst:215]: You should be using a `BatchSpanProcessor` here. `SimpleExportSpanProcessor` exports all spans sequentially and will be very slow.

File: google/cloud/pubsub_v1/publisher/client.py
Comment (chrisrossi) [inline google/cloud/pubsub_v1/publisher/client.py:383]: No space between "message" and "as".

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (cguardia) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:39]: There are no spaces at the end of these lines, which will cause words to be displayed together (e.g. couldnot, rather than could not).

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (cguardia) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:76]: Same issue from above, where spaces are missing at the end of the lines.

File: synth.py
Comment (cguardia) [inline synth.py:283]: Typo: dependencies.

File: google/cloud/pubsub_v1/opentelemetry_tracing.py
Comment (cguardia) [inline google/cloud/pubsub_v1/opentelemetry_tracing.py:40]: Just a suggestion: In addition to the import try, wouldn't it be possible to use an environment variable for explicitly disabling opentelemetry? That way the tests below would not need to mess with sys.modules.

**Review by cguardia — APPROVED** [review]
There are still a couple of suggestions from previous reviews, though they are mostly documentation and log warning formatting issues.

**Review by cguardia — CHANGES_REQUESTED** [review]
Seems like lint is not passing, plus there are a few failing tests. Please review.

**Comment by plamut** [comment]
Side note - before merging this (after releasing a new major version, of course), we must also run a few benchmarks to verify that the tracing does not hinder the performance, especially throughput. Mentioning as a a reminder to ourselves.

**Comment by johnbryan** [comment]
Hi all, Seth (author of this PR) finished his Google internship in mid-August so I assume is not actively maintaining this PR now. I would be willing to address the remaining comments here but will not be able to get to it for a couple of weeks. Seth if I'm mistaken and you are indeed still following/working on this let us know, but I assume you have other things to do!

**Comment by plamut** [comment]
@johnbryan That sounds good, thanks!

No rush, though, as we need too release a new major version first anyway.

**Comment by plamut** [comment]
@johnbryan FYI, a new major version has been released, unblocking this PR (which needs an update now, however).

**Comment by anguillanneuf** [comment]
@johnbryan Will you have a chance to rebase this PR?

**Comment by pamarc** [comment]
Hi, any news on this PR? This could be of use :)

**Comment by plamut** [comment]
Just as a quick status update, it is still not clear what exactly do want to be tracked, still awaiting the official specs. We can thus consider this PR blocked until further notice.

**Comment by CyberHippo** [comment]
Hi, is there any update/ETA on this PR?  :)

**Comment by dmallanb** [comment]
This PR has been open a while - any updates as to why it's still blocked?

**Review by pedroserrudo — APPROVED** [review]

File: google/cloud/pubsub_v1/publisher/client.py
Comment (dmaier-FN) [inline google/cloud/pubsub_v1/publisher/client.py:376]: What is the argument for doing this? It seems to me that, _best_ case, it results in the duplication of data in the message. But in the _worst_ case it actually leaks sensitive information into observability stack, just like logging the output of a SQL query would.

Comment (dmallanb) [reply google/cloud/pubsub_v1/publisher/client.py:376]: I've also noticed it's a potentially unneeded change after experimenting with alternatives myself. Trace information can be passed into pubsub attributes in the current iteration of the pubsub library using Open Telemetry manual instrumentation as far as I can see for now.

**Comment by mukund-ananthu** [comment]
Closing this, since its now done.

**Pull request closed without merging**
